staging: iio: adc: ad7192: Add low_pass_3db_filter_frequency

mirceacaprioru opened the pull request:
By adding this option we are able to remove the sync3 field and dt binding.
When setting the required cutoff frequency we also determine the ADC
configuration for chop and sync filter.

Signed-off-by: Mircea Caprioru <mircea.caprioru@analog.com>
(cherry picked from commit 5f3cbb78cd77e3af5971568746d4fc61eeae84f3)

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -534,6 +613,14 @@ static int ad7192_write_raw(struct iio_dev *indio_dev,
 		ad_sd_write_reg(&st->sd, AD7192_REG_MODE, 3, st->mode);
 		break;
+	case IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY:
+		if (!val) {
```
I don't think that you need this check

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -534,6 +613,14 @@ static int ad7192_write_raw(struct iio_dev *indio_dev,
+		}
+
+		ret = ad7192_set_3db_filter_freq(st, val);
+		break;
```
maybe return directly?

stefpopa replied:
makes sense

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -263,21 +267,12 @@ static int ad7192_setup(struct ad7192_state *st,
-	} else {
-		st->f_order = 1;
-	}
+	st->chop_en = false;
```
is `chop_en` still needed ? can we use `(st->conf & AD7192_CONF_CHOP)` to check the same ?

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -263,21 +267,12 @@ static int ad7192_setup(struct ad7192_state *st,
-	}
+	st->chop_en = false;
+	st->conf |= AD7192_CONF_CHOP;
+	st->f_order = AD7192_SYNC4_FILTER;
```
semantically, this changes now, as there is no `f_order == 1` anymore ;

commodo commented:
rest looks good

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -442,6 +464,63 @@ static unsigned int ad7192_get_temp_scale(bool unipolar)
+
+		st->chop_en = false;
+		st->conf &= ~AD7192_CONF_CHOP;
+		break;
```
do you need to sed `f_order` here in all switch cases ?

mirceacaprioru replied:
while following the comments from you i found a bug in the code and this need a bit of rework. will also try to simplify.

mirceacaprioru replied:
f_order is used in some other parts and by removing the chop_en field i have no indicator of what the order should be. It would be safer to set it each time the filter frequency changes.

stefpopa commented:
please add a change log every time when you force push a new commit

mirceacaprioru commented:
Changelog V1 -> V2
- removed chop_en
- modified filter frequency calculation
- moved filter frequency calculation to an inline function ad7192_get_available_filter_freqs()

mirceacaprioru commented:
Changelog V2 -> V3
- used DIV_ROUND_CLOSEST for divisions

mirceacaprioru commented:
Changelog V3 -> V4
-removed unnecessary brackets

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -406,6 +400,51 @@ static ssize_t ad7192_set(struct device *dev,
+static inline void ad7192_get_available_filter_freq(struct ad7192_state *st,
+						    int *freq)
+{
+	unsigned int fadc, temp[4];
```
you could remove the `temp` variable ; then you don't need the `memcpy()`

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -147,6 +147,10 @@
+#define AD7192_SYNC3_FILTER	3
+#define AD7192_SYNC4_FILTER	4
+#define AD7192_NO_SYNC_FILTER	1
+
```
nitpick: I would keep these in order:
```
#define AD7192_NO_SYNC_FILTER	1
#define AD7192_SYNC3_FILTER	3
#define AD7192_SYNC4_FILTER	4
```

stefpopa commented:
Looks good overall.

mirceacaprioru commented:
Changelog v4-> v5
- removed temp variable
- sorted sync filter defines

commodo commented:
[![LGTM](https://i.imgur.com/y1SlP0d.jpg)](https://i.imgur.com/y1SlP0d.jpg)

commodo commented:
Will merge tomorrow if no objections
